File: py/indicators/john_ehlers/corona/corona.py
"""Corona spectral analysis engine.

This is the shared engine consumed by CoronaSpectrum, CoronaSignalToNoiseRatio,
CoronaSwingPosition, and CoronaTrendVigor. It is NOT a registered indicator.
"""

import math
from dataclasses import dataclass, field
from typing import Optional

from .params import CoronaParams

# Constants matching Go implementation.
_HP_BUFFER_SIZE = 6
_FIR_COEF_SUM = 12.0

_DELTA_LOWER_THRESHOLD = 0.1
_DELTA_FACTOR = -0.015
_DELTA_SUMMAND = 0.5

_DC_BUFFER_SIZE = 5
_DC_MEDIAN_INDEX = 2

_DB_SMOOTHING_ALPHA = 0.33
_DB_SMOOTHING_ONE_MINUS = 0.67

_NORMALIZED_AMPLITUDE_FACTOR = 0.99
_DECIBELS_FLOOR = 0.01
_DECIBELS_GAIN = 10.0


@dataclass
class Filter:
    """Per-bin state of a single bandpass filter in the bank."""

    in_phase: float = 0.0
    in_phase_previous: float = 0.0
    quadrature: float = 0.0
    quadrature_previous: float = 0.0
    real: float = 0.0
    real_previous: float = 0.0
    imaginary: float = 0.0
    imaginary_previous: float = 0.0
    amplitude_squared: float = 0.0
    decibels: float = 0.0


class Corona:
    """Shared spectral-analysis engine for Ehlers Corona indicators.

    Call update(sample) once per bar. Read is_primed(), dominant_cycle,
    dominant_cycle_median, maximal_amplitude_squared, and filter_bank.
    """

    def __init__(self, params: Optional[CoronaParams] = None) -> None:
        cfg = CoronaParams() if params is None else CoronaParams(
            high_pass_filter_cutoff=params.high_pass_filter_cutoff,
            minimal_period=params.minimal_period,
            maximal_period=params.maximal_period,
            decibels_lower_threshold=params.decibels_lower_threshold,
            decibels_upper_threshold=params.decibels_upper_threshold,
        )
        _apply_defaults(cfg)
        _verify_parameters(cfg)

        self.minimal_period: int = cfg.minimal_period
        self.maximal_period: int = cfg.maximal_period
        self.minimal_period_times_two: int = cfg.minimal_period * 2
        self.maximal_period_times_two: int = cfg.maximal_period * 2
        self.decibels_lower_threshold: float = cfg.decibels_lower_threshold
        self.decibels_upper_threshold: float = cfg.decibels_upper_threshold

        self.filter_bank_length: int = self.maximal_period_times_two - self.minimal_period_times_two + 1
        self.filter_bank: list[Filter] = [Filter() for _ in range(self.filter_bank_length)]

        # Dominant cycle buffer initialized with MaxFloat64 sentinels.
        self._dc_buffer: list[float] = [math.inf] * _DC_BUFFER_SIZE
        self.dominant_cycle: float = math.inf
        self.dominant_cycle_median: float = math.inf

        # High-pass filter coefficients.
        phi = 2.0 * math.pi / cfg.high_pass_filter_cutoff
        self._alpha: float = (1.0 - math.sin(phi)) / math.cos(phi)
        self._half_one_plus_alpha: float = 0.5 * (1.0 + self._alpha)

        # Pre-calculated beta = cos(4π/n) for each half-period index.
        self._pre_calculated_beta: list[float] = []
        for index in range(self.filter_bank_length):
            n = self.minimal_period_times_two + index
            self._pre_calculated_beta.append(math.cos(4.0 * math.pi / n))

        # HP ring buffer.
        self._hp_buffer: list[float] = [0.0] * _HP_BUFFER_SIZE
        self._sample_previous: float = 0.0
        self._smooth_hp_previous: float = 0.0

        self.maximal_amplitude_squared: float = 0.0
        self._sample_count: int = 0
        self._primed: bool = False

    def is_primed(self) -> bool:
        """Whether the engine has seen enough samples for meaningful output."""
        return self._primed

    def update(self, sample: float) -> bool:
        """Feed the next sample. Returns True once primed."""
        if math.isnan(sample):
            return self._primed

        self._sample_count += 1

        # First sample: store as prior reference, no further processing.
        if self._sample_count == 1:
            self._sample_previous = sample
            return False

        # Step 1: High-pass filter.
        hp = self._alpha * self._hp_buffer[_HP_BUFFER_SIZE - 1] + \
            self._half_one_plus_alpha * (sample - self._sample_previous)
        self._sample_previous = sample

        # Shift buffer left.
        for i in range(_HP_BUFFER_SIZE - 1):
            self._hp_buffer[i] = self._hp_buffer[i + 1]
        self._hp_buffer[_HP_BUFFER_SIZE - 1] = hp

        # Step 2: 6-tap FIR smoothing {1,2,3,3,2,1}/12.
        smooth_hp = (self._hp_buffer[0] +
                     2 * self._hp_buffer[1] +
                     3 * self._hp_buffer[2] +
                     3 * self._hp_buffer[3] +
                     2 * self._hp_buffer[4] +
                     self._hp_buffer[5]) / _FIR_COEF_SUM

        # Step 3: Momentum.
        momentum = smooth_hp - self._smooth_hp_previous
        self._smooth_hp_previous = smooth_hp

        # Step 4: Adaptive delta.
        delta = _DELTA_FACTOR * self._sample_count + _DELTA_SUMMAND
        if delta < _DELTA_LOWER_THRESHOLD:
            delta = _DELTA_LOWER_THRESHOLD

        # Step 5: Filter-bank update.
        self.maximal_amplitude_squared = 0.0
        for index in range(self.filter_bank_length):
            n = self.minimal_period_times_two + index
            nf = float(n)

            gamma = 1.0 / math.cos(8.0 * math.pi * delta / nf)
            a = gamma - math.sqrt(gamma * gamma - 1.0)

            quadrature = momentum * (nf / (4.0 * math.pi))
            in_phase = smooth_hp

            half_one_min_a = 0.5 * (1.0 - a)
            beta = self._pre_calculated_beta[index]
            beta_one_plus_a = beta * (1.0 + a)

            f = self.filter_bank[index]

            real = half_one_min_a * (in_phase - f.in_phase_previous) + \
                beta_one_plus_a * f.real - a * f.real_previous
            imag = half_one_min_a * (quadrature - f.quadrature_previous) + \
                beta_one_plus_a * f.imaginary - a * f.imaginary_previous

            amp_sq = real * real + imag * imag

            f.in_phase_previous = f.in_phase
            f.in_phase = in_phase
            f.quadrature_previous = f.quadrature
            f.quadrature = quadrature
            f.real_previous = f.real
            f.real = real
            f.imaginary_previous = f.imaginary
            f.imaginary = imag
            f.amplitude_squared = amp_sq

            if amp_sq > self.maximal_amplitude_squared:
                self.maximal_amplitude_squared = amp_sq

        # Step 6: dB normalization and dominant-cycle weighted average.
        numerator = 0.0
        denominator = 0.0
        self.dominant_cycle = 0.0
        for index in range(self.filter_bank_length):
            f = self.filter_bank[index]

            decibels = 0.0
            if self.maximal_amplitude_squared > 0:
                normalized = f.amplitude_squared / self.maximal_amplitude_squared
                if normalized > 0:
                    arg = (1.0 - _NORMALIZED_AMPLITUDE_FACTOR * normalized) / _DECIBELS_FLOOR
                    if arg > 0:
                        decibels = _DECIBELS_GAIN * math.log10(arg)

            # EMA smoothing.
            decibels = _DB_SMOOTHING_ALPHA * decibels + _DB_SMOOTHING_ONE_MINUS * f.decibels
            if decibels > self.decibels_upper_threshold:
                decibels = self.decibels_upper_threshold
            f.decibels = decibels

            if decibels <= self.decibels_lower_threshold:
                n = float(self.minimal_period_times_two + index)
                adjusted = self.decibels_upper_threshold - decibels
                numerator += n * adjusted
                denominator += adjusted

        if denominator != 0:
            self.dominant_cycle = 0.5 * numerator / denominator
        if self.dominant_cycle < float(self.minimal_period):
            self.dominant_cycle = float(self.minimal_period)

        # Step 7: 5-sample median of dominant cycle.
        for i in range(_DC_BUFFER_SIZE - 1):
            self._dc_buffer[i] = self._dc_buffer[i + 1]
        self._dc_buffer[_DC_BUFFER_SIZE - 1] = self.dominant_cycle

        sorted_buf = sorted(self._dc_buffer)
        self.dominant_cycle_median = sorted_buf[_DC_MEDIAN_INDEX]
        if self.dominant_cycle_median < float(self.minimal_period):
            self.dominant_cycle_median = float(self.minimal_period)

        if self._sample_count < self.minimal_period_times_two:
            return False
        self._primed = True

        return True


def _apply_defaults(p: CoronaParams) -> None:
    """Fill zero/negative fields with Ehlers defaults."""
    if p.high_pass_filter_cutoff <= 0:
        p.high_pass_filter_cutoff = 30
    if p.minimal_period <= 0:
        p.minimal_period = 6
    if p.maximal_period <= 0:
        p.maximal_period = 30
    if p.decibels_lower_threshold == 0:
        p.decibels_lower_threshold = 6.0
    if p.decibels_upper_threshold == 0:
        p.decibels_upper_threshold = 20.0


def _verify_parameters(p: CoronaParams) -> None:
    """Validate parameters, raise ValueError on invalid config."""
    invalid = "invalid corona parameters"
    if p.high_pass_filter_cutoff < 2:
        raise ValueError(f"{invalid}: HighPassFilterCutoff should be >= 2")
    if p.minimal_period < 2:
        raise ValueError(f"{invalid}: MinimalPeriod should be >= 2")
    if p.maximal_period <= p.minimal_period:
        raise ValueError(f"{invalid}: MaximalPeriod should be > MinimalPeriod")
    if p.decibels_lower_threshold < 0:
        raise ValueError(f"{invalid}: DecibelsLowerThreshold should be >= 0")
    if p.decibels_upper_threshold <= p.decibels_lower_threshold:
        raise ValueError(f"{invalid}: DecibelsUpperThreshold should be > DecibelsLowerThreshold")

File: zig/src/indicators/john_ehlers/corona/corona.zig
const std = @import("std");
const math = std.math;

// Constants matching Go's corona package.
const default_high_pass_filter_cutoff = 30;
const default_minimal_period = 6;
const default_maximal_period = 30;
const default_decibels_lower_threshold = 6.0;
const default_decibels_upper_threshold = 20.0;

const high_pass_filter_buffer_size = 6;
const fir_coef_sum = 12.0;

const delta_lower_threshold = 0.1;
const delta_factor = -0.015;
const delta_summand = 0.5;

const dominant_cycle_buffer_size = 5;
const dominant_cycle_median_index = 2;

const decibels_smoothing_alpha = 0.33;
const decibels_smoothing_one_minus = 0.67;

const normalized_amplitude_factor = 0.99;
const decibels_floor = 0.01;
const decibels_gain = 10.0;

/// Per-bin state of a single bandpass filter in the bank.
pub const Filter = struct {
    in_phase: f64 = 0,
    in_phase_previous: f64 = 0,
    quadrature: f64 = 0,
    quadrature_previous: f64 = 0,
    real: f64 = 0,
    real_previous: f64 = 0,
    imaginary: f64 = 0,
    imaginary_previous: f64 = 0,
    amplitude_squared: f64 = 0,
    decibels: f64 = 0,
};

/// Parameters for creating a Corona engine.
pub const Params = struct {
    high_pass_filter_cutoff: i32 = 0,
    minimal_period: i32 = 0,
    maximal_period: i32 = 0,
    decibels_lower_threshold: f64 = 0,
    decibels_upper_threshold: f64 = 0,
};

pub const Error = error{
    InvalidHighPassFilterCutoff,
    InvalidMinimalPeriod,
    InvalidMaximalPeriod,
    InvalidDecibelsLowerThreshold,
    InvalidDecibelsUpperThreshold,
    OutOfMemory,
};

/// Corona is the shared spectral-analysis engine consumed by all corona indicators.
pub const Corona = struct {
    allocator: std.mem.Allocator,

    // Configuration (immutable after construction).
    minimal_period: i32,
    maximal_period: i32,
    minimal_period_times_two: i32,
    maximal_period_times_two: i32,
    filter_bank_length: usize,
    decibels_lower_threshold: f64,
    decibels_upper_threshold: f64,

    // High-pass filter coefficients.
    alpha: f64,
    half_one_plus_alpha: f64,

    // Pre-calculated cos(4π/n) for each half-period index.
    pre_calculated_beta: []f64,

    // HP ring buffer (oldest at index 0, current at index 5).
    high_pass_buffer: [high_pass_filter_buffer_size]f64,

    // Previous raw sample and previous smoothed HP.
    sample_previous: f64,
    smooth_hp_previous: f64,

    // Filter bank.
    filter_bank: []Filter,

    // Running maximum of amplitude-squared across filter bins for current bar.
    maximal_amplitude_squared: f64,

    // 5-sample ring buffer for dominant cycle median.
    dominant_cycle_buffer: [dominant_cycle_buffer_size]f64,

    // Sample counter.
    sample_count: i32,

    // Most recent dominant cycle estimate and its 5-sample median.
    dominant_cycle: f64,
    dominant_cycle_median: f64,

    primed: bool,

    pub fn init(allocator: std.mem.Allocator, p: ?Params) Error!Corona {
        var cfg = p orelse Params{};
        applyDefaults(&cfg);

        try verifyParameters(&cfg);

        const min_p = cfg.minimal_period;
        const max_p = cfg.maximal_period;
        const min_p2 = min_p * 2;
        const max_p2 = max_p * 2;
        const fbl: usize = @intCast(max_p2 - min_p2 + 1);

        // HP filter coefficients.
        const phi = 2.0 * math.pi / @as(f64, @floatFromInt(cfg.high_pass_filter_cutoff));
        const a = (1.0 - @sin(phi)) / @cos(phi);
        const half_one_plus_a = 0.5 * (1.0 + a);

        // Pre-calculate β = cos(4π / n) for each half-period index.
        const beta_buf = allocator.alloc(f64, fbl) catch return error.OutOfMemory;
        errdefer allocator.free(beta_buf);
        for (0..fbl) |index| {
            const n: f64 = @floatFromInt(min_p2 + @as(i32, @intCast(index)));
            beta_buf[index] = @cos(4.0 * math.pi / n);
        }

        const fb = allocator.alloc(Filter, fbl) catch return error.OutOfMemory;
        errdefer allocator.free(fb);
        @memset(fb, Filter{});

        var dc_buf: [dominant_cycle_buffer_size]f64 = undefined;
        @memset(&dc_buf, math.floatMax(f64));

        return .{
            .allocator = allocator,
            .minimal_period = min_p,
            .maximal_period = max_p,
            .minimal_period_times_two = min_p2,
            .maximal_period_times_two = max_p2,
            .filter_bank_length = fbl,
            .decibels_lower_threshold = cfg.decibels_lower_threshold,
            .decibels_upper_threshold = cfg.decibels_upper_threshold,
            .alpha = a,
            .half_one_plus_alpha = half_one_plus_a,
            .pre_calculated_beta = beta_buf,
            .high_pass_buffer = [_]f64{0} ** high_pass_filter_buffer_size,
            .sample_previous = 0,
            .smooth_hp_previous = 0,
            .filter_bank = fb,
            .maximal_amplitude_squared = 0,
            .dominant_cycle_buffer = dc_buf,
            .sample_count = 0,
            .dominant_cycle = math.floatMax(f64),
            .dominant_cycle_median = math.floatMax(f64),
            .primed = false,
        };
    }

    pub fn deinit(self: *Corona) void {
        self.allocator.free(self.pre_calculated_beta);
        self.allocator.free(self.filter_bank);
    }

    /// Feeds the next sample to the engine.
    /// Returns true once primed and outputs are meaningful.
    pub fn update(self: *Corona, sample: f64) bool {
        if (math.isNan(sample)) {
            return self.primed;
        }

        self.sample_count += 1;

        // First sample: just store as prior reference.
        if (self.sample_count == 1) {
            self.sample_previous = sample;
            return false;
        }

        // Step 1: High-pass filter.
        const hp = self.alpha * self.high_pass_buffer[high_pass_filter_buffer_size - 1] +
            self.half_one_plus_alpha * (sample - self.sample_previous);
        self.sample_previous = sample;

        // Shift buffer left.
        var i: usize = 0;
        while (i < high_pass_filter_buffer_size - 1) : (i += 1) {
            self.high_pass_buffer[i] = self.high_pass_buffer[i + 1];
        }
        self.high_pass_buffer[high_pass_filter_buffer_size - 1] = hp;

        // Step 2: 6-tap FIR smoothing {1, 2, 3, 3, 2, 1} / 12.
        const smooth_hp = (self.high_pass_buffer[0] +
            2.0 * self.high_pass_buffer[1] +
            3.0 * self.high_pass_buffer[2] +
            3.0 * self.high_pass_buffer[3] +
            2.0 * self.high_pass_buffer[4] +
            self.high_pass_buffer[5]) / fir_coef_sum;

        // Step 3: Momentum.
        const momentum = smooth_hp - self.smooth_hp_previous;
        self.smooth_hp_previous = smooth_hp;

        // Step 4: Adaptive delta.
        var delta = delta_factor * @as(f64, @floatFromInt(self.sample_count)) + delta_summand;
        if (delta < delta_lower_threshold) {
            delta = delta_lower_threshold;
        }

        // Step 5: Filter-bank update.
        self.maximal_amplitude_squared = 0;
        for (0..self.filter_bank_length) |index| {
            const n: f64 = @floatFromInt(self.minimal_period_times_two + @as(i32, @intCast(index)));

            const gamma = 1.0 / @cos(8.0 * math.pi * delta / n);
            const a_val = gamma - @sqrt(gamma * gamma - 1.0);

            const quadrature = momentum * (n / (4.0 * math.pi));
            const in_phase = smooth_hp;

            const half_one_min_a = 0.5 * (1.0 - a_val);
            const beta = self.pre_calculated_beta[index];
            const beta_one_plus_a = beta * (1.0 + a_val);

            var f = &self.filter_bank[index];

            const real = half_one_min_a * (in_phase - f.in_phase_previous) + beta_one_plus_a * f.real - a_val * f.real_previous;
            const imag = half_one_min_a * (quadrature - f.quadrature_previous) + beta_one_plus_a * f.imaginary - a_val * f.imaginary_previous;

            const amp_sq = real * real + imag * imag;

            f.in_phase_previous = f.in_phase;
            f.in_phase = in_phase;
            f.quadrature_previous = f.quadrature;
            f.quadrature = quadrature;
            f.real_previous = f.real;
            f.real = real;
            f.imaginary_previous = f.imaginary;
            f.imaginary = imag;
            f.amplitude_squared = amp_sq;

            if (amp_sq > self.maximal_amplitude_squared) {
                self.maximal_amplitude_squared = amp_sq;
            }
        }

        // Step 6: dB normalization and dominant-cycle weighted average.
        var numerator: f64 = 0;
        var denominator: f64 = 0;
        self.dominant_cycle = 0;

        for (0..self.filter_bank_length) |index| {
            var f = &self.filter_bank[index];

            var decibels: f64 = 0;
            if (self.maximal_amplitude_squared > 0) {
                const normalized = f.amplitude_squared / self.maximal_amplitude_squared;
                if (normalized > 0) {
                    const arg = (1.0 - normalized_amplitude_factor * normalized) / decibels_floor;
                    if (arg > 0) {
                        decibels = decibels_gain * @log10(arg);
                    }
                }
            }

            // EMA smoothing.
            decibels = decibels_smoothing_alpha * decibels + decibels_smoothing_one_minus * f.decibels;
            if (decibels > self.decibels_upper_threshold) {
                decibels = self.decibels_upper_threshold;
            }
            f.decibels = decibels;

            // Only bins at or below lower threshold contribute.
            if (decibels <= self.decibels_lower_threshold) {
                const n: f64 = @floatFromInt(self.minimal_period_times_two + @as(i32, @intCast(index)));
                const adjusted = self.decibels_upper_threshold - decibels;
                numerator += n * adjusted;
                denominator += adjusted;
            }
        }

        // DC = 0.5 * num / denom (converts half-period to period).
        if (denominator != 0) {
            self.dominant_cycle = 0.5 * numerator / denominator;
        }
        const min_period_f: f64 = @floatFromInt(self.minimal_period);
        if (self.dominant_cycle < min_period_f) {
            self.dominant_cycle = min_period_f;
        }

        // Step 7: 5-sample median of dominant cycle.
        {
            var j: usize = 0;
            while (j < dominant_cycle_buffer_size - 1) : (j += 1) {
                self.dominant_cycle_buffer[j] = self.dominant_cycle_buffer[j + 1];
            }
            self.dominant_cycle_buffer[dominant_cycle_buffer_size - 1] = self.dominant_cycle;
        }

        var sorted: [dominant_cycle_buffer_size]f64 = self.dominant_cycle_buffer;
        std.mem.sort(f64, &sorted, {}, std.sort.asc(f64));
        self.dominant_cycle_median = sorted[dominant_cycle_median_index];
        if (self.dominant_cycle_median < min_period_f) {
            self.dominant_cycle_median = min_period_f;
        }

        if (self.sample_count < self.minimal_period_times_two) {
            return false;
        }
        self.primed = true;

        return true;
    }

    // --- Accessors ---

    pub fn isPrimed(self: *const Corona) bool {
        return self.primed;
    }

    pub fn getDominantCycle(self: *const Corona) f64 {
        return self.dominant_cycle;
    }

    pub fn getDominantCycleMedian(self: *const Corona) f64 {
        return self.dominant_cycle_median;
    }

    pub fn getMaximalAmplitudeSquared(self: *const Corona) f64 {
        return self.maximal_amplitude_squared;
    }

    pub fn getFilterBank(self: *const Corona) []const Filter {
        return self.filter_bank;
    }

    pub fn getFilterBankLength(self: *const Corona) usize {
        return self.filter_bank_length;
    }

    pub fn getMinimalPeriod(self: *const Corona) i32 {
        return self.minimal_period;
    }

    pub fn getMaximalPeriod(self: *const Corona) i32 {
        return self.maximal_period;
    }

    pub fn getMinimalPeriodTimesTwo(self: *const Corona) i32 {
        return self.minimal_period_times_two;
    }

    pub fn getMaximalPeriodTimesTwo(self: *const Corona) i32 {
        return self.maximal_period_times_two;
    }
};

fn applyDefaults(p: *Params) void {
    if (p.high_pass_filter_cutoff <= 0) {
        p.high_pass_filter_cutoff = default_high_pass_filter_cutoff;
    }
    if (p.minimal_period <= 0) {
        p.minimal_period = default_minimal_period;
    }
    if (p.maximal_period <= 0) {
        p.maximal_period = default_maximal_period;
    }
    if (p.decibels_lower_threshold == 0) {
        p.decibels_lower_threshold = default_decibels_lower_threshold;
    }
    if (p.decibels_upper_threshold == 0) {
        p.decibels_upper_threshold = default_decibels_upper_threshold;
    }
}

fn verifyParameters(p: *const Params) Error!void {
    if (p.high_pass_filter_cutoff < 2) {
        return error.InvalidHighPassFilterCutoff;
    }
    if (p.minimal_period < 2) {
        return error.InvalidMinimalPeriod;
    }
    if (p.maximal_period <= p.minimal_period) {
        return error.InvalidMaximalPeriod;
    }
    if (p.decibels_lower_threshold < 0) {
        return error.InvalidDecibelsLowerThreshold;
    }
    if (p.decibels_upper_threshold <= p.decibels_lower_threshold) {
        return error.InvalidDecibelsUpperThreshold;
    }
}

// --- Tests ---
const testing = std.testing;
const testdata = @import("testdata.zig");


test "Corona default smoke" {
    var c = try Corona.init(testing.allocator, null);
    defer c.deinit();

    try testing.expectEqual(@as(usize, 49), c.getFilterBankLength());
    try testing.expectEqual(@as(i32, 12), c.getMinimalPeriodTimesTwo());
    try testing.expectEqual(@as(i32, 60), c.getMaximalPeriodTimesTwo());

    var primed_at: ?usize = null;
    for (0..testdata.test_input.len) |idx| {
        _ = c.update(testdata.test_input[idx]);
        if (c.isPrimed() and primed_at == null) {
            primed_at = idx;
        }
    }

    try testing.expect(primed_at != null);
    // Primes at sample index 11 (0-indexed), which is the 12th sample = MinimalPeriodTimesTwo.
    try testing.expectEqual(@as(usize, 11), primed_at.?);

    const dc = c.getDominantCycle();
    const dc_med = c.getDominantCycleMedian();
    const min_f: f64 = @floatFromInt(c.getMinimalPeriod());
    const max_f: f64 = @floatFromInt(c.getMaximalPeriod());

    try testing.expect(!math.isNan(dc) and !math.isInf(dc));
    try testing.expect(!math.isNan(dc_med) and !math.isInf(dc_med));
    try testing.expect(dc >= min_f and dc <= max_f);
    try testing.expect(dc_med >= min_f and dc_med <= max_f);

    const m = c.getMaximalAmplitudeSquared();
    try testing.expect(m > 0 and !math.isNan(m) and !math.isInf(m));
}

test "Corona NaN input is no-op" {
    var c = try Corona.init(testing.allocator, null);
    defer c.deinit();

    // Warm past priming.
    for (0..20) |idx| {
        _ = c.update(testdata.test_input[idx]);
    }
    try testing.expect(c.isPrimed());

    const dc_before = c.getDominantCycle();
    const dcm_before = c.getDominantCycleMedian();

    const got = c.update(math.nan(f64));
    try testing.expect(got); // preserves primed
    try testing.expectEqual(dc_before, c.getDominantCycle());
    try testing.expectEqual(dcm_before, c.getDominantCycleMedian());
}

test "Corona invalid params" {
    try testing.expectError(error.InvalidHighPassFilterCutoff, Corona.init(testing.allocator, .{ .high_pass_filter_cutoff = 1 }));
    try testing.expectError(error.InvalidMinimalPeriod, Corona.init(testing.allocator, .{ .minimal_period = 1 }));
    try testing.expectError(error.InvalidMaximalPeriod, Corona.init(testing.allocator, .{ .minimal_period = 10, .maximal_period = 10 }));
    try testing.expectError(error.InvalidDecibelsLowerThreshold, Corona.init(testing.allocator, .{ .decibels_lower_threshold = -1 }));
    try testing.expectError(error.InvalidDecibelsUpperThreshold, Corona.init(testing.allocator, .{ .decibels_lower_threshold = 6, .decibels_upper_threshold = 6 }));
}

test "Corona DC above min period" {
    var c = try Corona.init(testing.allocator, null);
    defer c.deinit();

    var saw_above_min = false;
    const min_f: f64 = @floatFromInt(c.getMinimalPeriod());

    for (0..testdata.test_input.len) |idx| {
        _ = c.update(testdata.test_input[idx]);
        if (c.isPrimed() and c.getDominantCycle() > min_f) {
            saw_above_min = true;
        }
    }

    try testing.expect(saw_above_min);
}
